### DISCUSSION TITLE
RFC: Lit Element

### DISCUSSION THREAD

**Discussion by DRiFTy17** [tyler-technologies-oss/forge#699] [RFCs]
The implementation philosophy for Forge has long been to stay as low-level and vanilla as possible when building Forge components. This direction was chosen to ensure the longevity of Forge, and reduce our dependency on other libraries or abstractions out there so we aren't at the mercy of their decision making and lifetime.

Forge was created in early 2017 when Web Components were still very new, and only supported in Chrome at the time. We took a risk by adopting this technology early on, relying on polyfills for the v1 custom elements spec in other browsers, and with the hope that these other browsers would catch up eventually (which of course they did very quickly). Looking back, we had a decision to make to hitch our wagon to an abstraction library such as Lit (which was still very early on its life at the time) or other tools like Stencil. These libraries both offered compelling reasons to use them, but at the time with everything being so new, we chose to stay vanilla and as close to the browser platform as possible.

We certainly do not regret this decision at all 7 years later, but it does the beg the question now that libraries like Lit Element are so prominent in the community with their many benefits, should we rethink this decisions and consider adopting [Lit](https://lit.dev/) for Forge?

## Why would we consider this now?

We still believe in our philosophy and stand behind our early decisions, but abstraction libraries such as Lit offer many benefits for a large UI library such as Forge. Additionally, with other projects in the pipeline such as Forge "extended" and other auxiliary libraries being based on Lit from the get go, it may be worth considering the added benefit of keeping the codebases in alignment by using Lit across the board.

## Benefits

- Smaller JavaScript bundle size.
  - Lit's abstraction and helpers allow for a much smaller production JavaScript output.
  - This also comes with the added benefit of reduced boilerplate code, which is something Forge struggles with in its current architecture.
- Increased performance.
  - Lit's lifecycle, directives, and template syntax are optimized for performance out of the box. This can be done with native DOM APIs, but it takes careful consideration and often times results in extra code.
- Lower barrier to entry for contributors
  - Lit is mature, well established, and has great documentation.
  - Components we build would be simpler to maintain, smaller, and potentially easier to understand with Lit.
- Declarative template syntax for rendering HTML.
  - This is where Lit really shines and produces lightweight and optimized components.
- Reactive primitives for optimized DOM updates.
- Common coding patterns and architecture across various Forge-based UI libraries.

## Drawbacks

- Hitching our wagon to something that may change or go away in the future causing us to rewrite/refactor code. We know how Google treats its tools...
- Requires learning Lit. You can no longer _just_ learn ~~basic HTML, JavaScript, and CSS~~ web standards to work on Forge.
  - On the flip side, Lit is quite easy to learn and has great documentation that doesn't stray _too_ far from the vanilla APIs.
  - I might argue that due to the Lit's abstraction, the learning curve to develop within Forge may actually be easier.
- Requires us to rewrite/refactor existing components.
  - This can be a gradual process though. Lit and vanilla-based components will be able to coexist while we migrate.
  -  We could even to choose to not refactor certain components at all.

## Alternatives?

There are other options out there that we could consider:

- Utilize just [lit-html](https://lit.dev/docs/v1/lit-html/introduction/) for rendering element templates.
  - This would reduce a lot of code on its own, and improve performance, but it also means we'd need to introduce our own reactivity model to Forge or manually re-render the templates when necessary.
- [Stencil](https://stenciljs.com/)
  - This is a mature Web Component compiler, but seems to have lost to Lit in popularity over the years...
- Re-architecture
  - We could consider adopting a new architecture and development practices in Forge and keep things vanilla.
  - This could mean we add our own abstraction layer and reactivity model on top of Forge (similar to what Lit already provides).
- Stay the course
  - Stick to our guns and keep things as they are now.
  - What we have works, and we can continue to improve it.

## What feedback are we looking for?

While it's not something we are planning to do in the near to medium term, we need group consensus about whether this is even worth considering for Tyler. Please provide any opinion you have on the subject both good and bad. We don't necessarily feel strongly one way or the other, but felt it was worth revisiting to gauge interest. This is also something we could progressively introduce over time as well.

**Comment by MikeMatusz** [comment]
I definitely like what I've seen of Lit so far.  It's extremely easy to follow the code, even reading through syntax I haven't seen before, it's usually intuitive what's going on.  I think Lit is a better fit than Stencil.

As far as the learning curve goes, I'm not sure "basic HTML, JavaScript, and CSS" adequately sums up the skillset required to be productive in the Forge codebase, I'd put it more on the "advanced" end of the spectrum.  I think the average dev probably needs to learn just as much about Shadow DOM and native browser APIs as they would about Lit, and the latter might actually be easier to pick up (though understanding the former would still be highly recommended).

I don't think maintaining our own abstraction layer is a good idea, as we don't have enough maintainers and most importantly, we'll never be able to compete with the support available for Lit questions/problems on StackOverflow and the wider internet.  My biggest concern would be what happens if Lit becomes no longer supported, but would the cost of having to migrate again outweigh the cost-savings of using the framework in the meantime?  Lots of variables there, but I feel like with its current trajectory, it will be around long enough for the math to come out in favor of using it.

Lastly, I would definitely prefer to see it introduced progressively.  I don't think getting every component migrated to Lit should be prioritized over other work.  Obviously we need to establish the pattern first with some existing components, and new components would be expected to use it, but I wouldn't want to see this delay other development.

**Reply by DRiFTy17** [reply]
Sorry, I probably over-generalized a bit too much when it comes to the "basic HTML, JavaScript, and CSS" point. I fully agree with you, and I intended to basically just convey that developers can learn and use just "web standards" when developing in Forge, rather than learning the syntax for a framework/library/abstraction is all. I see that as a "benefit" but at the same time I agree with you, learning Lit is actually probably easier and more familiar for new contributors and the code would be simpler and easier to follow most likely. I updated the original post, thanks for pointing that out!

Additionally, I also added another "alternative" regarding just the `lit-html` package on its own. This is something I considered in the past because we could easily adopt it for the rendering aspect without changing our core & adapter architecture at all. I'm not suggesting that this is _better_ than going all in on Lit Element, but just another option to consider.

I'm definitely on board with progressively introducing Lit if we do decide to move forward. Any new work could use it, and we could refactor existing components over time if we see the need, or if Lit could provide any obvious benefits that make it worth it if time allows.

**Comment by derekmoss** [comment]
I do think moving to Lit would greatly help us on the maintenance side as well as things like Stack overflow questions. I also agree with your sentiment about how those are google libraries and that is its own risk.

My thought is we already have a strong v3 that is complete and it works. If we did decide to move to Lit, and then 5 years down the road that turned out poorly because of the fears already mentioned. How hard would it be to pull v3 back up and update it? there would probably be some new components and some changed components. So that would be some amount of effort, and remembering our v3 architecture as that is a bit tribal atm.

I'm just wondering if the maintenance reduction and support reduction would be worth it even if we knew Lit might fail down the road. or if the pain of going back to our v3 architecture would be too great and cost us more in the long run. In which case we should avoid switching to Lit.

**Comment by Ross-Blakeney** [comment]
I have similar concerns to what's already been mentioned. Mainly: What do we do if/when lit disappears for whatever reason? How would we backtrack to where we are today and carry over all the delta? What does that look like?

I think we could pilot this with a small subset of components and see how we feel about it. We could take that a step further and really urge any teams asking for a bug fix or enhancement to do it themselves and see how that goes. I mean... I know that's our official stance already but I mean "really, really" make them do it.

**Comment by ryansheehan** [comment]
With respect to backing frameworks, Lit is an open source framework.  Just like any open source framework, should the project be pulled we have the option to find a new home, or fork Lit and take it on.

I don't have a strong opinion on this idea one way or the other.  If the development investment of hitching ourselves to Lit buys us reduced maintenance costs then it seems worth it.  If the effort to integrate Lit will not reduce our resources and efforts in maintaining the Forge, then it probably isn't worth it.  My assumption is that the effort is worth the investment since we would be tapping into a larger development effort that supports Lit.  

Do we need complete control of the underlying layers, or will adding Lit as a dependency alleviate maintenance needs?  How willing are we to contribute to the Lit project if there are gaps we need?

**Comment by DRiFTy17** [comment]
This was accepted and we have begun using it for new components, as well as started to refactor existing components. We'll continue to pick away at refactoring existing components as the year goes on.
